### PR TITLE
Optimize T5 for sequence generation

### PR DESCRIPTION
**Pull request by joecummings**
This PR makes the following changes to T5 to improve generation capabilities.
- Adds `prepare_inputs_for_generation` function to be compliant w/ `GenerationWrapper` API
- Adds `get_encoder` and `get_decoder` helper functions.
- Utilizes `past_key_values` to implement incremental decoding. This involves also a custom reorder cache function that can be used for beam search.
- Updates docstrings.
- Updates model weights to fit new architecture.
- Remove `T5Wrapper`
- Fix TorchScripting w/ new APIs

Testing:
- Passes existing tests
- Tests to come w/ generation integration tests (link)

Todo:
- [x] Add license to files
- [ ] Move to main folder
- [ ] Add model to README

### CODE REVIEW

**Review by mthrok — APPROVED** [review]
stamp

File: test/integration_tests/prototype/test_models.py [hunk 1]
```diff
@@ -79,7 +76,7 @@ def _t5_get_encoder(self, model, model_input, encoder_output):
         encoder = model.get_encoder()
         # Need to set the tgt_key_padding_mask to ensure the same results
         encoder_padding_mask = model_input.eq(model.padding_idx)
-        output_from_get_encoder = encoder(tgt=model_input, tgt_key_padding_mask=encoder_padding_mask)["encoder_output"]
+        output_from_get_encoder = encoder(model_input, src_key_padding_mask=encoder_padding_mask)["encoder_output"]
```
[the remark this answers is not in the document]
Comment (joecummings) [reply test/integration_tests/prototype/test_models.py:79]: changing the name of arguments, but yes this is prototype until tomorrow :)

File: torchtext/prototype/models/t5/model.py [hunk 2]
```diff
@@ -56,13 +55,13 @@ def __post_init__(self):
                 self.activation = "gelu_new"
 
 
-# NOTE: Comparable HuggingFace implentation can be found at https://github.com/huggingface/transformers/blob/8581a798c0a48fca07b29ce2ca2ef55adcae8c7e/src/transformers/models/t5/modeling_t5.py#L1269
```
[the remark this answers is not in the document]
Comment (joecummings) [reply torchtext/prototype/models/t5/model.py:59]: We include in the header that several functions are based on HF and I call it out in the docstring of those functions, as well. No need to say that there is a comparable HF implementation for ones that are just the normal Enc/Dec forward functions.

File: torchtext/prototype/models/t5/model.py [hunk 3]
```diff
@@ -160,14 +160,61 @@ def __init__(
...
+        # if decoder past is not included in output
+        # speedy decoding is disabled and no need to reorder
+        if past is None:
+            return past
+
+        reordered_decoder_past: List[Tuple[Tensor, Tensor, Tensor, Tensor]] = []
+        for layer_past_states in past:
+            # get the correct batch idx from layer past batch dim
+            # batch dim of `past` is at 2nd position
+            reordered_layer_past_states = ()
```
[the remark this answers is not in the document]
Comment (joecummings) [reply torchtext/prototype/models/t5/model.py:180]: yup :(

**Pull request merged**